### PR TITLE
CFY-4929 deployment update - phase I

### PR DESCRIPTION
**Pull request by AviaE**

### CODE REVIEW

File: rest-service/manager_rest/blueprints_manager.py [hunk 1]
```diff
@@ -503,14 +496,135 @@ def create_deployment(self, blueprint_id, deployment_id, inputs=None):
...
+                                          deployment_id,
+                                          dsl_node_instances):
+        node_instances = self.prepare_deployment_node_instances_for_storage(
+            deployment_id,
+            dsl_node_instances
+        )
+        self.store_deployment_node_instances(node_instances)
+
+    @staticmethod
+    def create_deployment_plan(blueprint_plan, inputs):
```
Comment (AviaE) [inline rest-service/manager_rest/blueprints_manager.py]: refactor create_deployment_plan and prepare_deployment_plan.
look at how it was before.

File: rest-service/manager_rest/blueprints_manager.py [hunk 2]
```diff
@@ -503,14 +496,135 @@ def create_deployment(self, blueprint_id, deployment_id, inputs=None):
...
+    def create_deployment(self, blueprint_id, deployment_id, inputs=None):
+
+        deployment_plan = self.prepare_deployment_plan(blueprint_id, inputs)
+
+        new_deployment = self.prepare_deployment_for_storage(
+            blueprint_id,
+            deployment_id,
+            deployment_plan,
+            inputs=inputs)
         self.sm.put_deployment(deployment_id, new_deployment)
```
Comment (AviaE) [inline rest-service/manager_rest/blueprints_manager.py]: make it a `store` method

File: rest-service/manager_rest/resources_v2_1.py [hunk 3]
```diff
@@ -276,6 +276,18 @@ def get(self, update_id):
             get_deployment_updates_manager().get_deployment_update(update_id)
 
 
+# Added to check full flow
```
Comment (AviaE) [inline rest-service/manager_rest/resources_v2_1.py]: remove

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+SOURCE_OPERATIONS = 'source_operations'
+TARGET_OPERATIONS = 'target_operations'
+PROPERTY = 'property'
+PROPERTIES = 'properties'
+WORKFLOW = 'workflow'
+WORKFLOWS = 'workflows'
+INTERFACES = 'interfaces'
+
+
+def _to_singular_form(plural_form):
```
Comment (AviaE) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: It is generally hated.

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        deployment = sm.get_deployment(deployment_id)
+        nodes = sm.get_nodes(
+            filters={'deployment_id': [deployment_id]}).items
+
+        return cls(deployment, nodes)
+
+    @classmethod
+    def from_deployment_update(cls, deployment_update):
+
+        blueprints_manager = BlueprintsManager()
```
Comment (AviaE) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: use get_blueprints_manager.
it also has get_storage_manager

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
+#########
```
Comment (AviaE) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: Underscores when needed (which is most of the time)

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+
+        self._extract_steps(new, old)
+        self.flip_diff_perspective()
+        self._extract_steps(old, new)
+
+    def _get_matching_relationship(self, relationship, relationships):
+        relationship_type = relationship[TYPE]
+        relationship_target = relationship[TARGET_ID]
+        for other_relationship in relationships:
+            if other_relationship.get(TYPE, '') == relationship_type and \
```
Comment (AviaE) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: remove the `''` and fix the corresponding tests

File: rest-service/manager_rest/blueprints_manager.py [hunk 8]
```diff
@@ -503,14 +496,135 @@ def create_deployment(self, blueprint_id, deployment_id, inputs=None):
...
+                relationships=self._prepare_node_relationships(raw_node)
+            ))
+        return nodes
+
+    def store_deployment_nodes(self, nodes):
+        for node in nodes:
+            self.sm.put_node(node)
+
+    @staticmethod
+    def prepare_deployment_node_instances_for_storage(deployment_id,
```
Comment (AviaE) [inline rest-service/manager_rest/blueprints_manager.py]: No sad smileys

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+                del self.entity_id[-1]
+
+        del self.entity_id[-1]
+
+    def _extract_steps_from_nodes(self,
+                                  new_nodes,
+                                  old_nodes
+                                  ):
+        self.entity_id.append(NODES)
+        for node_name in new_nodes:
```
Comment (AviaE) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: x, y in iteritems

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+        # plan, but you only need to create one step of modification.
+        # We chose, arbitrarily, to create the modification step when we
+        # compare the new to the old (_flip_diff_perspective == False).
+        elif modify and self._flip_diff_perspective:
+            return
+
+        elif not self._flip_diff_perspective:
+            self.deployment_update_manager.create_deployment_update_step(
+                self.deployment_update_id,
+                'add',
```
Comment (AviaE) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: 'add' if self._flip_diff_perspective else 'remove'

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+
+    def _extract_steps_from_operations(self,
+                                       operations_type_name,
+                                       operations,
+                                       old_operations,
+                                       ):
+        adjusted_operations_type_name = \
+            self._transform_interfaces_name_to_operation_name(
+                operations_type_name
+            )
```
Comment (AviaE) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: Up =(

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+            return name.replace(INTERFACES, OPERATIONS)
+        return name
+
+    def _extract_steps_from_operations(self,
+                                       operations_type_name,
+                                       operations,
+                                       old_operations,
+                                       ):
+        adjusted_operations_type_name = \
+            self._transform_interfaces_name_to_operation_name(
```
Comment (AviaE) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: Enquire...
including the `operations_type_name` naming

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+                operations_type_name
+            )
+        self.entity_id.append(adjusted_operations_type_name)
+        for operation_name in operations:
+            if not DeploymentUpdateStepsExtractor._is_operation_copy(
```
Comment (AviaE) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: don't ignore the copies.

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+
+            elif entities_name == WORKFLOWS:
+                self._extract_steps_from_entities(
+                    WORKFLOWS, new_entities, old_entities)
+
+    def _create_deployment_update_step(self, entity_type, modify=False):
+
+        entity_id_string = str(self.entity_id)
+
+        if modify and not self._flip_diff_perspective:
```
Comment (AviaE) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: move the second condition to its own line

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+class DeploymentUpdateStepsExtractor(object):
+
+    def __init__(self, deployment_update):
+        self.deployment_id = deployment_update.deployment_id
+        self.deployment_update_id = deployment_update.id
+        self.deployment_update_manager = \
+            manager_rest.deployment_update.manager.\
+            get_deployment_updates_manager()
+        self.sm = self.deployment_update_manager.sm
+        self.entity_id = EntityID()
```
Comment (ran-z) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: entity_id_holder?
entity_id_builder?
current_entity_id?

entity_id --> entity_path?

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+        self.old_deployment_plan = None
+        self.new_deployment_plan = None
+        self._flip_diff_perspective = False
```
Comment (ran-z) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: change name

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 17]
```diff
@@ -0,0 +1,306 @@
...
+                        if (relationship[PROPERTIES] !=
+                                matching_relationship[PROPERTIES]):
+                            # alert the user that changing the relationship
+                            # is currently not supported during deployment
+                            # updates.
+                            pass
+            else:
+                self._create_deployment_update_step(RELATIONSHIP)
+
+            del self.entity_id[-1]
```
Comment (ran-z) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: belongs in the EntityId class; should use a context manager for this

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+        self.deployment_update_manager = \
+            manager_rest.deployment_update.manager.\
+            get_deployment_updates_manager()
+        self.sm = self.deployment_update_manager.sm
+        self.entity_id = EntityID()
+        self.old_deployment_plan = None
+        self.new_deployment_plan = None
+        self._flip_diff_perspective = False
+
+    def flip_diff_perspective(self):
```
Comment (ran-z) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: unnecessairy

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+    steps_extractor = \
+        DeploymentUpdateStepsExtractor(deployment_update)
+
+    # update the steps extractor with the old and the new deployment plans
+    steps_extractor.old_deployment_plan = \
+        DeploymentPlan.from_storage(steps_extractor.deployment_id)
+    steps_extractor.new_deployment_plan = \
+        DeploymentPlan.from_deployment_update(deployment_update)
+
+    steps_extractor.extract_steps()
```
Comment (ran-z) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: i dont like the side affects on this; i think i'd either extract all steps then do the actual changes here, or even at the caller's side

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 20]
```diff
@@ -0,0 +1,306 @@
...
+
+            elif entities_name == OUTPUTS:
+                self._extract_steps_from_entities(
+                    OUTPUTS, new_entities, old_entities)
+
+            elif entities_name == WORKFLOWS:
+                self._extract_steps_from_entities(
+                    WORKFLOWS, new_entities, old_entities)
+
+    def _create_deployment_update_step(self, entity_type, modify=False):
```
Comment (AviaE) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: A wrapper function that captures the logic of determining the action. If there is an action, just call _create_deployment_update_step (that just calls create_deployment_update_step)

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 21]
```diff
@@ -119,14 +119,32 @@ def __init__(self, deployment_update):
...
+
+        # The following property determines the type of steps that will be
+        # created as part of a call to self._extract_steps.
+
+        # If it is False, the diff process (which is composed of two calls to
+        # self.extract_steps) will detect entities that were not in
+        # `old_deployment_plan` but exist in `new_deployment_plan`.
+        # That is, the diff will detect all the entities that were added as
+        # part of the deployment update process.
+        # In addition, if the diff perspective is not inversed, the diff will
```
Comment (AviaE) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: Furthermore

File: rest-service/manager_rest/test/test_deployment_update_step_extraction.py [hunk 22]
```diff
@@ -0,0 +1,1096 @@
+import copy
+import json
+import mock
```
Comment (AviaE) [inline rest-service/manager_rest/test/test_deployment_update_step_extraction.py]: add line

File: rest-service/manager_rest/test/test_deployment_update_step_extraction.py [hunk 23]
```diff
@@ -0,0 +1,1096 @@
...
+        names_to_mock = [
+            'manager_rest.deployment_update.step_extractor.'
+            'manager_rest.deployment_update.manager'
+        ]
+        for name_to_mock in names_to_mock:
+            patcher = patch(name_to_mock)
+            self.addCleanup(patcher.stop)
+            patcher.start()
+
+        self.stub_deployment_update = models.DeploymentUpdate(
```
Comment (AviaE) [inline rest-service/manager_rest/test/test_deployment_update_step_extraction.py]: no self.

File: rest-service/manager_rest/test/test_deployment_update_step_extraction.py [hunk 24]
```diff
@@ -0,0 +1,1081 @@
...
+            NODE,
+            'nodes:node1')
+
+        create_step = \
+            self.deployment_update_manager.create_deployment_update_step
+
+        create_step_calls = create_step.call_args_list
+        self.assertTrue(add_node_call in create_step_calls)
+        self.assertTrue(remove_node_call in create_step_calls)
+        self.assertEquals(len(create_step_calls), 2)
```
Comment (AviaE) [inline rest-service/manager_rest/test/test_deployment_update_step_extraction.py]: All the expected value are on the left

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 25]
```diff
@@ -0,0 +1,360 @@
...
+
+        nodes = blueprints_manager.prepare_deployment_nodes_for_storage(
+            blueprint_id,
+            deployment_id,
+            deployment_plan
+        )
+        return cls(deployment, nodes)
+
+    @staticmethod
+    def _transform_nodes(nodes):
```
Comment (AviaE) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: Add test

File: rest-service/manager_rest/deployment_update/step_extractor.py [hunk 26]
```diff
@@ -0,0 +1,360 @@
...
+
+            elif entities_name == OUTPUTS:
+                self._extract_steps_from_entities(
+                    OUTPUTS, new_entities, old_entities)
+
+            elif entities_name == WORKFLOWS:
+                self._extract_steps_from_entities(
+                    WORKFLOWS, new_entities, old_entities)
+
+    def determine_step_action(self, modify):
```
Comment (AviaE) [inline rest-service/manager_rest/deployment_update/step_extractor.py]: add test

File: tests/workflow_tests/test_deployment_update.py [hunk 27]
```diff
@@ -203,19 +203,15 @@ def _test_add_node(self, archive_mode=False):
         try:
```
Comment (AviaE) [inline tests/workflow_tests/test_deployment_update.py]: No need to call it twice. In addition, use keyword when called

**Comment by AviaE** [comment]
reviewed by @ran-z

**Pull request merged**